test/integration/component/test_escalations_templates.py

```diff
@@ -915,3 +968,223 @@ def test_04_copy_template(self):
+            except CloudstackAPIException  as e:
+                self.assertRaises("Exception Raised : %s" % e)
+
+            admin_vm_created = VirtualMachine.create(
+                self.admin_api_client,
+                self.services["virtual_machine"],
+                accountid=admin_account.name,
+                domainid=admin_account.domainid,
+                serviceofferingid=self.service_offering.id,
+            )
+            self.assertIsNotNone(
+                admin_vm_created,
+                "VM creation failed"
+            )
+            # Get the Root disk of VM
+            volume = list_volumes(
+                self.admin_api_client,
+                virtualmachineid=admin_vm_created.id,
+                type='ROOT',
+                listall=True
+            )
+            snapshot_created = Snapshot.create(
+                self.admin_api_client,
+                volume[0].id,
+                account=admin_account.name,
+                domainid=admin_account.domainid
+            )
+            self.assertIsNotNone(
+                snapshot_created,
+                "Snapshot creation failed"
+            )
+            self.debug("Creating a template from snapshot: %s" % snapshot_created.id)
+            #
+            #    Generate public template from the snapshot
+            try:
+                admin_template = Template.create_from_snapshot(
+                    self.admin_api_client,
+                    snapshot_created,
+                    self.services["template"]
+                )
+                self.updateConfigurAndRestart("allow.public.user.templates", "true")
+                self.fail("Template creation passed from snapshot for domain admin")
+            except CloudstackAPIException  as e:
+                self.assertRaises("Exception Raised : %s" % e)
+
+            VirtualMachine.stop(admin_vm_created, self.admin_api_client)
+            list_stopped_vms_after = VirtualMachine.list(
+                self.admin_api_client,
+                listall=self.services["listall"],
+                domainid=admin_account.domainid,
+                state="Stopped")
+            status = validateList(list_stopped_vms_after)
+            self.assertEquals(
+                PASS,
+                status[0],
+                "Stopped VM is not in Stopped state"
+            )
+            try:
+                admin_template = Template.create(
+                    self.admin_api_client, self.services["template"],
+                    volume[0].id
+                )
+                self.updateConfigurAndRestart("allow.public.user.templates", "true")
+                self.fail("Template creation passed from volume for domain admin")
+            except CloudstackAPIException  as e:
+                self.assertRaises("Exception Raised : %s" % e)
+
+            self.updateConfigurAndRestart("allow.public.user.templates", "true")
```

Comment:
Have this step in tearDownClass() also. In case test case fails before executing completely, the config value will remain to False. We need to explicitly ensure it is set to true even if test case fails at any stage.